aiida/tools/visualization/graph.py

```diff
@@ -0,0 +1,740 @@
+        :type style: dict or None
+        :param overwrite: whether to overrite existing edge (Default value = False)
+        :type overwrite: bool
+
+        """
+        in_node = self._load_node(in_node)
+        if in_node.pk not in self._nodes:
+            raise AssertionError("the in_node must have already been added to the graph")
+        out_node = self._load_node(out_node)
+        if out_node.pk not in self._nodes:
+            raise AssertionError("the out_node must have already been added to the graph")
+
+        if (in_node.pk, out_node.pk, link_pair) in self.edges and not overwrite:
+            return
+
+        style = {} if style is None else style
+        self._edges.add((in_node.pk, out_node.pk, link_pair))
+        style.update(self._global_edge_style)
+
+        _add_graphviz_edge(self._graph, in_node, out_node, style)
+
+    def add_incoming(self, node, link_types=(), annotate_links=False, return_pks=True):
+        """add nodes and edges for incoming links to a node
+
+        :param node: node or node pk/uuid
+        :type node: aiida.orm.nodes.node.Node or int
+        :param link_types: filter by link types (Default value = ())
+        :type link_types: str or tuple[str] or aiida.common.links.LinkType or tuple[aiida.common.links.LinkType]
+        :param annotate_links: label edges with the link 'label', 'type' or 'both' (Default value = False)
+        :type annotate_links: bool or str
+        :param return_pks: whether to return a list of nodes, or list of node pks (Default value = True)
+        :type return_pks: bool
+        :returns: list of nodes or node pks
+
+        """
+        if annotate_links not in [False, "label", "type", "both"]:
+            raise AssertionError('annotate_links must be one of False, "label", "type" or "both"')
+
+        if link_types:
+            if isinstance(link_types, six.string_types):
+                link_types = [link_types]
+            link_types = tuple(
+                [getattr(LinkType, l.upper()) if isinstance(l, six.string_types) else l for l in link_types])
+
+        nodes = []
+        for link_triple in node.get_incoming(link_type=link_types).link_triples:
+            self.add_node(link_triple.node)
+            link_pair = LinkPair(link_triple.link_type, link_triple.link_label)
+            style = self._link_styles(link_triple.link_type)
+            if annotate_links == "label":
+                style['label'] = link_triple.link_label
+            elif annotate_links == "type":
+                style['label'] = link_triple.link_type.name
+            elif annotate_links == "both":
+                style['label'] = "{}\n{}".format(link_triple.link_type.name, link_triple.link_label)
+            self.add_edge(link_triple.node, node, link_pair, style=style)
+            nodes.append(link_triple.node.pk if return_pks else link_triple.node)
+
+        return nodes
+
+    def add_outgoing(self, node, link_types=(), annotate_links=False, return_pks=True):
```

Comment:
this function shares quite a bit of code and interface with `add_incoming`.
Do we really need two functions here?